packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1309,63 +1307,63 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
+        return plugin.isDirectDependency;
+      });
+      if (directDependencies.isNotEmpty) {
+        if (directDependencies.length > 1) {
+          globals.printError(
+              'Plugin ${entry.key}:$platformKey has conflicting direct dependency implementations:\n'
+                  '${directDependencies.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
+                  'To fix this issue, remove all but one of these dependencies from pubspec.yaml.\n'
+          );
+          hasResolutionError = true;
+        } else {
+          pluginResolution.add(directDependencies.first);
+        }
+        continue;
+      }
+      // Next, defer to the default implementation if there is one.
+      final String? defaultPackageName = defaultImplementations[entry.key];
+      if (defaultPackageName != null) {
+        final int defaultIndex = candidates
+            .indexWhere((Plugin plugin) => plugin.name == defaultPackageName);
+        if (defaultIndex != -1) {
+          pluginResolution.add(candidates[defaultIndex]);
+          continue;
+        }
+      }
+      // Otherwise, require an explicit choice.
+      if (candidates.length > 1) {
         globals.printError(
-          'Plugin ${entry.key} has conflicting direct dependency implementations:\n'
-          '${directDependencies.map((PluginInterfaceResolution r) => '  ${r.plugin.name}\n').join()}'
-          'To fix this issue, remove all but one of these dependencies from pubspec.yaml.\n'
+            'Plugin ${entry.key}:$platformKey has multiple possible implementations:\n'
+                '${candidates.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
+                'To fix this issue, add one of these dependencies to pubspec.yaml.\n'
         );
         hasResolutionError = true;
-      } else {
-        finalResolution.add(directDependencies.first);
-      }
-      continue;
-    }
-    // Next, defer to the default implementation if there is one.
-    final String? defaultPackageName = defaultImplementations[entry.key];
-    if (defaultPackageName != null) {
-      final int defaultIndex = candidates
-          .indexWhere((PluginInterfaceResolution r) => r.plugin.name == defaultPackageName);
-      if (defaultIndex != -1) {
-        finalResolution.add(candidates[defaultIndex]);
         continue;
       }
     }
-    // Otherwise, require an explicit choice.
-    if (candidates.length > 1) {
-      globals.printError(
-        'Plugin ${entry.key} has multiple possible implementations:\n'
-        '${candidates.map((PluginInterfaceResolution r) => '  ${r.plugin.name}\n').join()}'
-        'To fix this issue, add one of these dependencies to pubspec.yaml.\n'
-      );
-      hasResolutionError = true;
-      continue;
-    }
+
+    finalResolution.addAll(
+      pluginResolution.map((Plugin plugin) =>
+          PluginInterfaceResolution(plugin: plugin, platform: platformKey)),
+    );
```

Comment:
Also the resolution can be handled per platform (for the same reason with the unique keys). And is needed to be able to extract this logic _per platform_ in a separate method later to reuse it for the resolution of native plugins.